pkg/policy/l4_filter_deny_test.go

```diff
@@ -0,0 +1,676 @@
+	c.Assert(state.matchedDenyRules, Equals, 1)
+	resDeny.Detach(testSelectorCache)
+	expected.Detach(testSelectorCache)
+
+	state = traceState{}
+	resDeny, err = shadowRule.resolveIngressPolicy(testPolicyContext, toFoo, &state, L4PolicyMap{}, nil, nil)
+	c.Assert(err, IsNil)
+	c.Assert(resDeny, IsNil)
+	c.Assert(state.selectedRules, Equals, 0)
+	c.Assert(state.matchedRules, Equals, 0)
+	c.Assert(state.matchedDenyRules, Equals, 0)
+}
+
+// Case 3: deny all on L7 in both rules, but select different endpoints in each rule.
+func (ds *PolicyTestSuite) TestMergingWithDifferentEndpointSelectedDenyAllL7(c *C) {
+
+	selectDifferentEndpointsDenyAllL7 := &rule{
+		Rule: api.Rule{
+			EndpointSelector: endpointSelectorA,
+			IngressDeny: []api.IngressDenyRule{
+				{
+					IngressCommonRule: api.IngressCommonRule{
+						FromEndpoints: []api.EndpointSelector{endpointSelectorA},
+					},
+					ToPorts: []api.PortDenyRule{{
+						Ports: []api.PortProtocol{
+							{Port: "80", Protocol: api.ProtoTCP},
+						},
+					}},
+				},
+				{
+					IngressCommonRule: api.IngressCommonRule{
+						FromEndpoints: []api.EndpointSelector{endpointSelectorC},
+					},
+					ToPorts: []api.PortDenyRule{{
+						Ports: []api.PortProtocol{
+							{Port: "80", Protocol: api.ProtoTCP},
+						},
+					}},
+				},
+			},
+		}}
+
+	buffer := new(bytes.Buffer)
+	ctxToA := SearchContext{To: labelsA, Trace: TRACE_VERBOSE}
+	ctxToA.Logging = stdlog.New(buffer, "", 0)
+	c.Log(buffer)
+
+	expected := L4PolicyMap{"80/TCP": &L4Filter{
+		Port:     80,
+		Protocol: api.ProtoTCP,
+		U8Proto:  6,
+		wildcard: nil,
+		L7Parser: ParserTypeNone,
+		L7RulesPerSelector: L7DataMap{
+			cachedSelectorA: &PerSelectorPolicy{IsDeny: true},
+			cachedSelectorC: &PerSelectorPolicy{IsDeny: true},
+		},
+		Ingress:          true,
+		DerivedFromRules: labels.LabelArrayList{nil, nil},
+	}}
+
+	state := traceState{}
+	resDeny, err := selectDifferentEndpointsDenyAllL7.resolveIngressPolicy(testPolicyContext, &ctxToA, &state, L4PolicyMap{}, nil, nil)
+	c.Assert(err, IsNil)
+	c.Assert(resDeny, Not(IsNil))
+	c.Assert(resDeny, checker.DeepEquals, expected)
+	c.Assert(state.selectedRules, Equals, 1)
+	c.Assert(state.matchedRules, Equals, 0)
+	c.Assert(state.matchedDenyRules, Equals, 1)
+	resDeny.Detach(testSelectorCache)
+	expected.Detach(testSelectorCache)
+
+	buffer = new(bytes.Buffer)
+	ctxToC := SearchContext{To: labelsC, Trace: TRACE_VERBOSE}
+	ctxToC.Logging = stdlog.New(buffer, "", 0)
+	c.Log(buffer)
+
+	state = traceState{}
+	resDeny, err = selectDifferentEndpointsDenyAllL7.resolveIngressPolicy(testPolicyContext, toFoo, &state, L4PolicyMap{}, nil, nil)
+	c.Assert(err, IsNil)
+	c.Assert(resDeny, IsNil)
+	c.Assert(state.selectedRules, Equals, 0)
+	c.Assert(state.matchedRules, Equals, 0)
+	c.Assert(state.matchedDenyRules, Equals, 0)
+}
+
+// Case 4: allow all at L3/L7 in one rule, and deny a selected an endpoint in
+// another rule. Should resolve to just allowing all on L3/L7 (first rule
```

Comment:
`L3/L4`, not `L3/L7`

Reply:
-> `deny a selected endpoint`

Reply:
Yes, I need to take a 2nd look at all test case descriptions.